torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -164,16 +176,26 @@ at::Tensor DoBinaryOp(const at::Tensor& self, const at::Scalar& other,
   return bridge::AtenFromXlaTensor(result);
 }
 
-void CheckBinaryOpTypePromotion(const at::Tensor& out, const at::Tensor& self,
-                                const at::Tensor& other) {
-  at::ScalarType resultType = at::result_type(self, other);
-  XLA_CHECK(at::canCast(/*from=*/resultType, /*to=*/out.scalar_type()));
+template <typename B>
+at::Tensor DoBinaryOpWithoutPromo(const at::Tensor& self,
+                                  const at::Tensor& other, const B& bin_op) {
+  at::ScalarType dtype = at::result_type(self, other);
+  XLA_CHECK(at::canCast(/*from=*/dtype, /*to=*/self.scalar_type()));
+  std::pair<XLATensor, XLATensor> operands =
+      GetBinaryOperands(self, UnwrapNumber(other, dtype));
+  XLATensor result = bin_op(operands.first, operands.second);
+  return bridge::AtenFromXlaTensor(result);
 }
 
-void CheckBinaryOpTypePromotion(const at::Tensor& out, const at::Tensor& self,
-                                const at::Scalar& other) {
-  at::ScalarType resultType = at::result_type(self, other);
-  XLA_CHECK(at::canCast(/*from=*/resultType, /*to=*/out.scalar_type()));
+template <typename B>
+void DoBinaryOpOut(const at::Tensor& self, const at::Tensor& other,
+                   at::Tensor& out, const B& bin_op_out) {
+  at::ScalarType dtype = at::result_type(self, other);
+  XLA_CHECK(at::canCast(/*from=*/dtype, /*to=*/out.scalar_type()));
```

Comment:
in contrast, this looks reasonable, because the computed intermediate promotion result then has to cast to out